File: src_py/prompt_guard/benchmark.py
from __future__ import annotations

import argparse
import json
import time
from pathlib import Path
from typing import Dict, Iterable, List, Tuple

from .config import ScanConfig
from .core import scan_text
from .fixtures import BENCHMARK_CASES, BenchmarkCase, load_benchmark_cases


def run_benchmark(cases: Iterable[BenchmarkCase] = BENCHMARK_CASES) -> Dict[str, object]:
    case_results = []
    totals = {
        "prompt_injection": _empty_metrics(),
        "pii": _empty_metrics(),
        "contradictory": _empty_metrics(),
    }
    family_totals: Dict[str, Tuple[int, int]] = {}
    latencies_ms: List[float] = []

    for case in cases:
        case_config = ScanConfig.with_rules(
            rule_packs=case.rule_packs,
            pii_locales=case.pii_locales,
            scan_decoded_payloads=case.scan_decoded_payloads,
        )
        started = time.perf_counter()
        report = scan_text(case.text, config=case_config)
        latencies_ms.append((time.perf_counter() - started) * 1000.0)
        prompt_hit = report.is_prompt_injection == case.prompt_injection
        pii_hit = report.has_pii == case.pii
        contradiction_hit = report.is_contradictory == case.contradictory

        _record(totals["prompt_injection"], report.is_prompt_injection, case.prompt_injection)
        _record(totals["pii"], report.has_pii, case.pii)
        _record(totals["contradictory"], report.is_contradictory, case.contradictory)

        if case.prompt_injection:
            correct, total = family_totals.get(case.family, (0, 0))
            family_totals[case.family] = (correct + int(report.is_prompt_injection), total + 1)

        case_results.append(
            {
                "id": case.id,
                "family": case.family,
                "expected": {
                    "prompt_injection": case.prompt_injection,
                    "pii": case.pii,
                    "contradictory": case.contradictory,
                },
                "actual": {
                    "prompt_injection": report.is_prompt_injection,
                    "pii": report.has_pii,
                    "contradictory": report.is_contradictory,
                },
                "scores": {
                    "prompt_injection": report.prompt_injection_score,
                    "pii": report.pii_score,
                    "contradictory": report.contradiction_score,
                },
                "top_prompt_injection_rules": [finding.label for finding in report.prompt_injection_findings[:3]],
            }
        )

    return {
        "headline": _headline(family_totals),
        "families": {
            family: {"detected": correct, "total": total, "rate": _rate(correct, total)}
            for family, (correct, total) in sorted(family_totals.items())
        },
        "tasks": {
            name: _summarize_metrics(values)
            for name, values in totals.items()
        },
        "latency": _summarize_latency(latencies_ms),
        "cases": case_results,
    }


def _summarize_latency(latencies_ms: List[float]) -> Dict[str, float]:
    if not latencies_ms:
        return {"scans": 0, "mean_ms": 0.0, "p95_ms": 0.0, "max_ms": 0.0}
    ordered = sorted(latencies_ms)
    p95_index = min(len(ordered) - 1, int(round(0.95 * (len(ordered) - 1))))
    return {
        "scans": len(ordered),
        "mean_ms": round(sum(ordered) / len(ordered), 4),
        "p95_ms": round(ordered[p95_index], 4),
        "max_ms": round(ordered[-1], 4),
    }


def main() -> None:
    parser = argparse.ArgumentParser(description="Run PromptGuard benchmark fixtures.")
    parser.add_argument("--cases", type=Path, help="Path to a JSON benchmark fixture file.")
    parser.add_argument("--json", action="store_true", help="Print JSON only.")
    args = parser.parse_args()
    cases = load_benchmark_cases(args.cases) if args.cases else BENCHMARK_CASES
    result = run_benchmark(cases)
    if args.json:
        print(json.dumps(result, indent=2, sort_keys=True))
        return
    print(format_benchmark(result))


def format_benchmark(result: Dict[str, object]) -> str:
    lines = [str(result["headline"]), ""]
    lines.append("Attack-family detection:")
    for family, stats in result["families"].items():  # type: ignore[union-attr]
        lines.append(f"- {family}: {stats['rate']:.1%} ({stats['detected']}/{stats['total']})")
    lines.append("")
    lines.append("Task accuracy:")
    for task, stats in result["tasks"].items():  # type: ignore[union-attr]
        if task == "contradictory":
            task = "contradictory (experimental)"
        lines.append(f"- {task}: {stats['accuracy']:.1%} ({stats['correct']}/{stats['total']})")
    lines.append("")
    latency = result["latency"]  # type: ignore[index]
    lines.append("Scan latency:")
    lines.append(
        f"- {latency['scans']} scans: mean {latency['mean_ms']:.3f} ms, "
        f"p95 {latency['p95_ms']:.3f} ms, max {latency['max_ms']:.3f} ms"
    )
    lines.append("")
    lines.append("JSON:")
    lines.append(json.dumps(result, indent=2, sort_keys=True))
    return "\n".join(lines)


def _headline(family_totals: Dict[str, Tuple[int, int]]) -> str:
    minja = family_totals.get("MINJA", (0, 0))
    return f"Included MINJA fixture detection: {minja[0]}/{minja[1]} synthetic cases"


def _empty_metrics() -> Dict[str, int]:
    return {"tp": 0, "tn": 0, "fp": 0, "fn": 0}


def _record(values: Dict[str, int], actual: bool, expected: bool) -> None:
    if actual and expected:
        values["tp"] += 1
    elif actual and not expected:
        values["fp"] += 1
    elif not actual and expected:
        values["fn"] += 1
    else:
        values["tn"] += 1


def _summarize_metrics(values: Dict[str, int]) -> Dict[str, object]:
    correct = values["tp"] + values["tn"]
    total = sum(values.values())
    precision = _rate(values["tp"], values["tp"] + values["fp"])
    recall = _rate(values["tp"], values["tp"] + values["fn"])
    f1 = _rate(2 * precision * recall, precision + recall) if precision + recall else 0.0
    return {
        "correct": correct,
        "total": total,
        "accuracy": _rate(correct, total),
        "precision": precision,
        "recall": recall,
        "f1": f1,
        "confusion": dict(values),
    }


def _rate(correct: float, total: float) -> float:
    if total == 0:
        return 0.0
    return correct / total


if __name__ == "__main__":
    main()

File: src_ts/benchmark.ts
import { scanText } from "./index.js";
import { BENCHMARK_CASES, loadBenchmarkCases, type BenchmarkCase } from "./fixtures.js";

export { loadBenchmarkCases, BENCHMARK_CASES, type BenchmarkCase } from "./fixtures.js";

interface Count {
  correct: number;
  total: number;
}

interface Metrics {
  tp: number;
  tn: number;
  fp: number;
  fn: number;
}

export function runBenchmark(cases: BenchmarkCase[] = BENCHMARK_CASES): Record<string, unknown> {
  const tasks: Record<string, Metrics> = {
    prompt_injection: emptyMetrics(),
    pii: emptyMetrics(),
    contradictory: emptyMetrics(),
  };
  const families = new Map<string, Count>();
  const caseResults = [];
  const latenciesMs: number[] = [];

  for (const item of cases) {
    const started = performance.now();
    const report = scanText(item.text, {
      config: {
        rulePacks: item.rulePacks ?? ["core"],
        piiLocales: item.piiLocales ?? ["us"],
        scanDecodedPayloads: item.scanDecodedPayloads ?? false,
      },
    });
    latenciesMs.push(performance.now() - started);
    const expected = {
      prompt_injection: item.promptInjection ?? false,
      pii: item.pii ?? false,
      contradictory: item.contradictory ?? false,
    };
    const actual = {
      prompt_injection: report.isPromptInjection,
      pii: report.hasPii,
      contradictory: report.isContradictory,
    };

    record(tasks.prompt_injection, actual.prompt_injection, expected.prompt_injection);
    record(tasks.pii, actual.pii, expected.pii);
    record(tasks.contradictory, actual.contradictory, expected.contradictory);

    if (expected.prompt_injection) {
      const count = families.get(item.family) ?? { correct: 0, total: 0 };
      count.correct += Number(actual.prompt_injection);
      count.total += 1;
      families.set(item.family, count);
    }

    caseResults.push({
      id: item.id,
      family: item.family,
      expected,
      actual,
      scores: {
        prompt_injection: report.promptInjectionScore,
        pii: report.piiScore,
        contradictory: report.contradictionScore,
      },
      top_prompt_injection_rules: report.promptInjectionFindings.slice(0, 3).map((finding) => finding.label),
    });
  }

  const familyStats = Object.fromEntries(
    [...families.entries()]
      .sort(([left], [right]) => left.localeCompare(right))
      .map(([family, count]) => [family, { detected: count.correct, total: count.total, rate: rate(count) }]),
  );

  return {
    headline: headline(families),
    families: familyStats,
    tasks: Object.fromEntries(
      Object.entries(tasks).map(([name, count]) => [name, summarizeMetrics(count)]),
    ),
    latency: summarizeLatency(latenciesMs),
    cases: caseResults,
  };
}

function summarizeLatency(latenciesMs: number[]): Record<string, number> {
  if (latenciesMs.length === 0) return { scans: 0, mean_ms: 0, p95_ms: 0, max_ms: 0 };
  const ordered = [...latenciesMs].sort((a, b) => a - b);
  const p95Index = Math.min(ordered.length - 1, Math.round(0.95 * (ordered.length - 1)));
  const mean = ordered.reduce((sum, value) => sum + value, 0) / ordered.length;
  const round4 = (value: number) => Math.round(value * 10000) / 10000;
  return {
    scans: ordered.length,
    mean_ms: round4(mean),
    p95_ms: round4(ordered[p95Index]),
    max_ms: round4(ordered[ordered.length - 1]),
  };
}

export function formatBenchmark(result: Record<string, unknown>): string {
  const lines = [String(result.headline), "", "Attack-family detection:"];
  const families = result.families as Record<string, { detected: number; total: number; rate: number }>;
  for (const [family, stats] of Object.entries(families)) {
    lines.push(`- ${family}: ${(stats.rate * 100).toFixed(1)}% (${stats.detected}/${stats.total})`);
  }
  lines.push("", "Task accuracy:");
  const tasks = result.tasks as Record<string, { correct: number; total: number; accuracy: number }>;
  for (const [task, stats] of Object.entries(tasks)) {
    const label = task === "contradictory" ? "contradictory (experimental)" : task;
    lines.push(`- ${label}: ${(stats.accuracy * 100).toFixed(1)}% (${stats.correct}/${stats.total})`);
  }
  const latency = result.latency as { scans: number; mean_ms: number; p95_ms: number; max_ms: number };
  lines.push(
    "",
    "Scan latency:",
    `- ${latency.scans} scans: mean ${latency.mean_ms.toFixed(3)} ms, ` +
      `p95 ${latency.p95_ms.toFixed(3)} ms, max ${latency.max_ms.toFixed(3)} ms`,
  );
  lines.push("", "JSON:", JSON.stringify(result, null, 2));
  return lines.join("\n");
}

function emptyMetrics(): Metrics {
  return { tp: 0, tn: 0, fp: 0, fn: 0 };
}

function record(count: Metrics, actual: boolean, expected: boolean): void {
  if (actual && expected) count.tp += 1;
  else if (actual && !expected) count.fp += 1;
  else if (!actual && expected) count.fn += 1;
  else count.tn += 1;
}

function rate(count: Count): number {
  return count.total === 0 ? 0 : count.correct / count.total;
}

function divide(numerator: number, denominator: number): number {
  return denominator === 0 ? 0 : numerator / denominator;
}

function summarizeMetrics(count: Metrics): Record<string, unknown> {
  const correct = count.tp + count.tn;
  const total = count.tp + count.tn + count.fp + count.fn;
  const precision = divide(count.tp, count.tp + count.fp);
  const recall = divide(count.tp, count.tp + count.fn);
  const f1 = precision + recall === 0 ? 0 : (2 * precision * recall) / (precision + recall);
  return {
    correct,
    total,
    accuracy: divide(correct, total),
    precision,
    recall,
    f1,
    confusion: { ...count },
  };
}

function headline(families: Map<string, Count>): string {
  const minja = families.get("MINJA") ?? { correct: 0, total: 0 };
  return `Included MINJA fixture detection: ${minja.correct}/${minja.total} synthetic cases`;
}

if (import.meta.url === `file://${process.argv[1]}`) {
  const casesArgIndex = process.argv.indexOf("--cases");
  const jsonOnly = process.argv.includes("--json");
  const cases =
    casesArgIndex >= 0 && process.argv[casesArgIndex + 1]
      ? loadBenchmarkCases(process.argv[casesArgIndex + 1])
      : BENCHMARK_CASES;
  const result = runBenchmark(cases);
  console.log(jsonOnly ? JSON.stringify(result, null, 2) : formatBenchmark(result));
}
